[WK2] Add and use TriviallyCopyableArgumentCoder

zdobersek opened the pull request:
#### c71a406e692818f81b1d341f9dcfe4acdea039d2
<pre>
[WK2] Add and use TriviallyCopyableArgumentCoder
<a href="https://bugs.webkit.org/show_bug.cgi?id=251511">https://bugs.webkit.org/show_bug.cgi?id=251511</a>

Reviewed by NOBODY (OOPS!).

Add TriviallyCopyableArgumentCoder, a simple struct template providing encode
and decode functions for any trivially copyable class that invoke encodeObject()
and decodeObject() functions on the encoder and decoder classes, respectively.

Different ArgumentCoder specializations (assuming they are specialized for some
trivially copyable type) can inherit from this struct to gain the necessary
encoding and decoding functionality.

ArgumentCoder specialization for any arithmetic type is changed to inherit from
TriviallyCopyableArgumentCoder. ArgumentCoder specializations inheriting from
this struct are added for different types currently using encodeObject() and
decodeObject() methods directly. AudioStreamBasicDescription specialization
avoids inheritance so that the Cocoa-specific CAAudioStreamDescription header
is limited to being included only in the WebCoreArgumentCoders implementation
file.

SimpleArgumentCoder usage and inheritance is removed in favor of using
TriviallyCopyableArgumentCoder.

* Source/WTF/wtf/ArgumentCoder.h:
(IPC::TriviallyCopyableArgumentCoder::encode):
(IPC::TriviallyCopyableArgumentCoder::decode):
(IPC::ArgumentCoder&lt;bool&gt;::encode):
(IPC::ArgumentCoder&lt;bool&gt;::decode):
* Source/WebCore/platform/audio/cocoa/CAAudioStreamDescription.h:
(WebCore::CAAudioStreamDescription::encode const):
(WebCore::CAAudioStreamDescription::decode):
* Source/WebCore/platform/mediastream/RealtimeMediaSourceCapabilities.h:
(WebCore::CapabilityValueOrRange::encode const):
(WebCore::CapabilityValueOrRange::decode):
* Source/WebKit/Platform/IPC/ArgumentCoders.h:
(IPC::SimpleArgumentCoder::encode): Deleted.
(IPC::SimpleArgumentCoder::decode): Deleted.
* Source/WebKit/Shared/TextCheckerState.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.cpp:
(IPC::ArgumentCoder&lt;AudioStreamBasicDescription&gt;::encode):
(IPC::ArgumentCoder&lt;AudioStreamBasicDescription&gt;::decode):
(IPC::ArgumentCoder&lt;RectEdges&lt;bool&gt;&gt;::encode): Deleted.
(IPC::ArgumentCoder&lt;RectEdges&lt;bool&gt;&gt;::decode): Deleted.
(IPC::ArgumentCoder&lt;ViewportArguments&gt;::encode): Deleted.
(IPC::ArgumentCoder&lt;ViewportArguments&gt;::decode): Deleted.
* Source/WebKit/Shared/WebCoreArgumentCoders.h:
* Source/WebKit/Shared/win/WebCoreArgumentCodersWin.cpp:
(IPC::ArgumentCoder&lt;LOGFONT&gt;::encode): Deleted.
(IPC::ArgumentCoder&lt;LOGFONT&gt;::decode): Deleted.
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c71a406e692818f81b1d341f9dcfe4acdea039d2

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/106010 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/15065 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/38842 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/115193 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/175301 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/109914 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/16489 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/6248 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/98219 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/114929 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/111763 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/12554 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/95524 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/40094 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/94419 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/27167 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/81763 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/95668 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/8321 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/28519 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/95002 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/6134 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/8811 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/5080 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/30490 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/14435 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/48062 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/103744 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/10355 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/25719 "Passed tests") | 
| | | | | 
<!--EWS-Status-Bubble-End-->

kkinnunen-apple commented on Source/WebCore/platform/audio/cocoa/CAAudioStreamDescription.h:
```diff
@@ -99,3 +100,7 @@ inline CAAudioStreamDescription toCAAudioStreamDescription(const AudioStreamDesc
 
 }
+
+namespace IPC {
```
To layer the code correctly, the direct references to IPC namespace must happen only from WebKit level.

This means that this should go int WebCoreArgumentCodersCocoa.h or WebCoreArgumentCoders.h in cocoa-specific ifdef

I know currently there are the IPC friend declarations, those should eventually be removed..
Also likely ArgumentCoder.h was moved into wrong place, WTF. Hopefully these can be moved back.

kkinnunen-apple commented on Source/WebCore/platform/mediastream/RealtimeMediaSourceCapabilities.h:
```diff
@@ -277,4 +278,8 @@ template<> struct EnumTraits<WebCore::CapabilityValueOrRange::Type> {
 
 } // namespace WTF
 
+namespace IPC {
```
same with this

zdobersek committed:
[WK2] Add and use TriviallyCopyableArgumentCoder
https://bugs.webkit.org/show_bug.cgi?id=251511

Reviewed by NOBODY (OOPS!).

Add TriviallyCopyableArgumentCoder, a simple struct template providing encode
and decode functions for any trivially copyable class that invoke encodeObject()
and decodeObject() functions on the encoder and decoder classes, respectively.

Different ArgumentCoder specializations (assuming they are specialized for some
trivially copyable type) can inherit from this struct to gain the necessary
encoding and decoding functionality.

ArgumentCoder specialization for any arithmetic type is changed to inherit from
TriviallyCopyableArgumentCoder. ArgumentCoder specializations inheriting from
this struct are added for different types currently using encodeObject() and
decodeObject() methods directly. AudioStreamBasicDescription specialization
avoids inheritance so that the Cocoa-specific CAAudioStreamDescription header
is limited to being included only in the WebCoreArgumentCoders implementation
file.

SimpleArgumentCoder usage and inheritance is removed in favor of using
TriviallyCopyableArgumentCoder.

* Source/WTF/wtf/ArgumentCoder.h:
(IPC::TriviallyCopyableArgumentCoder::encode):
(IPC::TriviallyCopyableArgumentCoder::decode):
(IPC::ArgumentCoder<bool>::encode):
(IPC::ArgumentCoder<bool>::decode):
* Source/WebCore/platform/audio/cocoa/CAAudioStreamDescription.h:
(WebCore::CAAudioStreamDescription::encode const):
(WebCore::CAAudioStreamDescription::decode):
* Source/WebCore/platform/mediastream/RealtimeMediaSourceCapabilities.h:
(WebCore::CapabilityValueOrRange::encode const):
(WebCore::CapabilityValueOrRange::decode):
* Source/WebKit/Platform/IPC/ArgumentCoders.h:
(IPC::SimpleArgumentCoder::encode): Deleted.
(IPC::SimpleArgumentCoder::decode): Deleted.
* Source/WebKit/Shared/TextCheckerState.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.cpp:
(IPC::ArgumentCoder<AudioStreamBasicDescription>::encode):
(IPC::ArgumentCoder<AudioStreamBasicDescription>::decode):
(IPC::ArgumentCoder<RectEdges<bool>>::encode): Deleted.
(IPC::ArgumentCoder<RectEdges<bool>>::decode): Deleted.
(IPC::ArgumentCoder<ViewportArguments>::encode): Deleted.
(IPC::ArgumentCoder<ViewportArguments>::decode): Deleted.
* Source/WebKit/Shared/WebCoreArgumentCoders.h:
* Source/WebKit/Shared/win/WebCoreArgumentCodersWin.cpp:
(IPC::ArgumentCoder<LOGFONT>::encode): Deleted.
(IPC::ArgumentCoder<LOGFONT>::decode): Deleted.
